src/observer.jl

```diff
@@ -1,72 +1,117 @@
 """
-TomographyObserver is an implementation of an
+Observer is an implementation of an
 observer object (<:AbstractObserver) which
-implements custom measurements.
+implements custom measurements to perform
+at each layer of the circuit evolution.
 """
-struct TomographyObserver <: AbstractObserver
-  fidelity::Vector{Float64}
-  train_loss::Vector{Float64}
-  test_loss::Vector{Float64}
-  trace_preserving_distance::Vector{Float64}
-  frobenius_distance::Vector{Float64}
-  fidelity_bound::Vector{Float64}
-  function TomographyObserver()
-    return new([],[],[],[],[],[])
-  end
+struct Observer <: AbstractObserver
+  measurements::Dict{String, Pair{<:Union{Nothing,Function, String, Tuple, Pair{<:Function, <:Any}},<:Any}}
 end
 
-train_loss(obs::TomographyObserver) = obs.train_loss
-test_loss(obs::TomographyObserver) = obs.test_loss
-trace_preserving_distance(obs::TomographyObserver) = 
-  obs.trace_preserving_distance
-fidelity(o::TomographyObserver) = o.fidelity
-fidelity_bound(o::TomographyObserver) = o.fidelity_bound
-frobenius_distance(o::TomographyObserver) = o.frobenius_distance
-
-function measure!(obs::TomographyObserver;
-                  train_loss=nothing,
-                  test_loss=nothing,
-                  trace_preserving_dist=nothing,
-                  frob_dist=nothing,
-                  F=nothing,Fbound=nothing)
-  
-  ## Record negative log likelihood
-  if !isnothing(train_loss)
-    push!(obs.train_loss,train_loss)
-  end
-  if !isnothing(test_loss)
-    push!(obs.test_loss,test_loss)
-  end
-  # Measure fidelity
-  if !isnothing(trace_preserving_dist)
-    push!(trace_preserving_distance(obs),trace_preserving_dist)
-  end
-  # Measure fidelity
-  if !isnothing(F)
-    push!(fidelity(obs),F)
-  end
-  # Measure fidelity
-  if !isnothing(frob_dist)
-    push!(frobenius_distance(obs),frob_dist)
+Observer() = Observer(Dict{String, Pair{<:Union{Nothing,Function, String, Tuple, Pair{<:Function, <:Any}},<:Any}}())
+
+observable(observer::Observer, observable::String) = first(observer.measurements[observable])
```

Comment:
Maybe use a more generic term like `measurement`, since it may not actually be an observable (for example the norm or the entropy).